fv3gfs/util/halo_updater.py

```diff
@@ -0,0 +1,263 @@
+        cls,
+        comm: "Communicator",
+        numpy_like_module: NumpyModule,
+        specifications: Iterable[HaloUpdateSpec],
+        boundaries: Iterable[Boundary],
+        tag: int,
+        optional_timer: Optional[Timer] = None,
+        finalize_on_wait: bool = False,
+    ) -> "HaloUpdater":
+        """Create/retrieve as many packed buffer as needed and queue the slices to exchange.
+
+        Args:
+            comm: communicator to post network messages
+            numpy_like_module: module implementing numpy API
+            specifications: data specifications to exchange, including number of halo points
+            boundaries: informations on the exchange boundaries.
+            tag: network tag (to differentiate messaging) for this node.
+            optional_timer: timing of operations.
+
+        Returns:
+            HaloUpdater ready to exchange data.
+        """
+
+        timer = optional_timer if optional_timer is not None else NullTimer()
+
+        exchange_descriptors = defaultdict(list)
+        for boundary in boundaries:
+            for specification in specifications:
+                exchange_descriptors[boundary.to_rank].append(
+                    HaloExchangeData(
+                        specification,
+                        boundary.send_slice(specification),
+                        boundary.n_clockwise_rotations,
+                        boundary.recv_slice(specification),
+                    )
+                )
+
+        transformers = {}
+        for rank, exchange_descriptor in exchange_descriptors.items():
+            transformers[rank] = HaloDataTransformer.get(
+                numpy_like_module, exchange_descriptor
+            )
+
+        return cls(comm, tag, transformers, timer, finalize_on_wait)
+
+    @classmethod
+    def from_vector_specifications(
+        cls,
+        comm: "Communicator",
+        numpy_like_module: NumpyModule,
+        specifications_x: Iterable[HaloUpdateSpec],
+        specifications_y: Iterable[HaloUpdateSpec],
+        boundaries: Iterable[Boundary],
+        tag: int,
+        optional_timer: Optional[Timer] = None,
+        finalize_on_wait: bool = False,
+    ) -> "HaloUpdater":
+        """Create/retrieve as many packed buffer as needed and queue the slices to exchange.
+
+        Args:
+            comm: communicator to post network messages
+            numpy_like_module: module implementing numpy API
+            specifications_x: specifications to exchange along the x axis.
+                          Length must match y specifications.
+            specifications_y: specifications to exchange along the y axis.
+                          Length must match x specifications.
+            boundaries: informations on the exchange boundaries.
+            tag: network tag (to differentiate messaging) for this node.
+            n_halo_points: size of the halo to exchange.
+            optional_timer: timing of operations.
```

Comment:
Please remove n_halo_points and add finalize_on_wait to docstring.

Reply:
Removed.